plugins/offboard/offboard.h

```diff
@@ -152,6 +157,13 @@ class Offboard : public PluginBase
      */
     void set_velocity_body(VelocityBodyYawspeed velocity_body_yawspeed);
 
+    /**
+     * @brief Set the altitude.
+     *
+     * @param altitude Altitude 'float'.
+     */
+    void set_altitude(Altitude atlitude);
```

Comment:
I wonder if there shouldn't be just an interface for position which then includes altitude, instead of altitude alone.

Reply:
I chose it to be separate since sometimes you'd want to only specify an altitude and ignore x and y position. In our specific case, we want to specify an attitude together with an altitude.

Reply:
Ok, do you want to specify an attitude or a velocity? Because an attitude leads to different velocities due to wind or airframe control imbalances.

Reply:
I need to specify an attitude, no velocity. I am aware that this will result in different velocities for different circumstances. I will most likely do this through the set_attitude MAVLINK command.

Reply:
May I ask why you need attitude commands and not velocity commands? DroneCore aims to be high-level and based on autonomous capabilites of drones, away from rate or attitude control.

Reply:
>> I wonder if there shouldn't be just an interface for position which then includes altitude, instead of altitude alone.

>   @xgerrmann I chose it to be separate since sometimes you'd want to only specify an altitude and ignore x and y position. In our specific case, we want to specify an attitude together with an altitude.

FWIW I think that we do want a generic position controller - someone is going to want that, and then we'd need to change the API. There is nothing to stop us accepting NAN (or similar) for parameters that should default to current values.

Reply:
@hamishwillee yes I agree. But this here is essentially "altitude control".

Reply:
I understand that also a position control is required for dronecore, e.g. offboard->set_position(x,y,z). The set_altitude functionality should not be a replacement for this. I see offboard control a way to do the more 'advanced' things, like the set_altitude while ignoring position x,y. In the MAVLINK specification it is possible to set ignore flags for specific control targets and I think this functionality should be covered.
As I understood from what hamishwillee suggested, we can also make a set_position(x,y,z) where either x,y,z or a combination of those can be set to null (null=ignore). To me this gives the functionality I need without being too specific. Does this sound like a decent solution to you guys @julianoes @hamishwillee ?

Reply:
Yes, I think that's a solution, however, we would use `NaN` instead of null because you might want go to position `0` 😄.

I need to also look into the changes that this requires on the Firmware side to see the implications.

Reply:
How is the MAVLINK message SET_POSITION_TARGET_LOCAL_NED interpreted and how is it different from SET_POSITION_TARGET_GLOBAL_INT (which is not yet implemented on the px4 side btw) ? The LOCAL NED, is this always with respect to home, of always with respect to the current position?

Reply:
I don't actually know how and if this is all implemented. I suggest feel free to dig into the offboard control code and find what is happening and what is missing. That would certainly be a great!

Reply:
> `set_position(x,y,z) where either x,y,z or a combination of these can be set to null (null=ignore). To me this gives the functionality I need without being too specific. Does this sound like a decent solution to you guys @julianoes @hamishwillee ?

Yes, my suggestion is  `set_position(x,y,z)` with a `NaN` for ignored values (ie "use current"). 

> How is the MAVLINK message SET_POSITION_TARGET_LOCAL_NED interpreted and how is it different from SET_POSITION_TARGET_GLOBAL_INT (which is not yet implemented on the px4 side btw) ? The LOCAL NED, is this always with respect to home, of always with respect to the current position?

My recollection is that SET_POSITION_TARGET_GLOBAL_INT is global GPS co-ordinates, [SET_POSITION_TARGET_LOCAL_NED](http://mavlink.org/messages/common#SET_POSITION_TARGET_LOCAL_NED) is local co-cordinates where you specify the frame- e.g.  MAV_FRAME_LOCAL_NED is relative to home location. There are a number of different frames in the standard but I don't know which ones PX4 supports.

Reply:
A further complication is that the global position altitude can be relative to MSL or to home.  I'd make the API relative to home. 
Don't know what the NED API should look like yet - depends a bit on what frames PX4 supports.

Reply:
> Yes, my suggestion is set_position(x,y,z) with a NaN for ignored values (ie "use current").
So this is a crucial point. Here I would really ignore x and y and set the corresponding ignore_position flag, since the user does not specify them and apparently does not care.

Reply:
@xgerrmann yes if we set `NaN` in the DroneCore API, then it should set the ignore flags to true for these fields. 

Can you make a PR for the change needed on the Firmware side? If you haven't already.

Reply:
Update: I will continue on this PR soon (1-2 days).